### DISCUSSION TITLE
Should we use `Buffer` or `TypedArrays` when working with binary?

### DISCUSSION THREAD

**Discussion by larsqa** [nodejs/help#4792] [Q&A]
> Prior to the introduction of `TypedArray` in ECMAScript 2015 (ES6), the JavaScript language had no mechanism for reading or manipulating streams of binary data. The `Buffer` class was introduced as part of the Node.js API to make it possible to interact with octet streams in the context of things like TCP streams and file system operations.

However, after [typed arrays were introduced with ES6 (in 2015)](https://262.ecma-international.org/6.0/#sec-typedarray-objects) standard, the Node.js `Buffer` was re-implemented to extend from `Uint8Array` for better compatibility and performance. This made `Buffer` a kind of typed array.

Several online sources, including StackOverflow answers, mention that `Buffer` has a bigger API and better performance, compared to ES6 `TypedArrays`. 

It's obvious that `Buffer` has a larger API, where `Buffer` seems to have [approximately 62 methods ](https://nodejs.org/api/buffer.html#class-buffer), while `TypedArray` have only [approximately 34 ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray).

However, I'm unable to find anything trustworthy regarding performance?  Previous [documentation version mentioned that `Buffer.slices` is more efficient than `TypedArray.slices`](https://nodejs.org/docs/latest-v9.x/api/buffer.html#buffer_buffers_and_typedarray), but that statement has since [been removed in recent versions.](https://nodejs.org/api/buffer.html#buffers-and-typedarrays)

Furthermore, the [Node.JS docs recommend using `TypedArray`'s like `Uint8Array` with `.slices()` since the `Buffer`'s implementation can have unexpected behaviour.](https://nodejs.org/api/buffer.html#buffers-and-typedarrays)

So, is `Buffer` mainly a good choice because of it's larger API, or, are there still some performance related features around it?

**Comment by bnoordhuis** [comment]
Depends on what you're doing. Uint8Array is fine for simple byte manipulations but you probably can't beat buffer.includes() with your lovingly open-coded version except for trivial searches.

**Comment by SaiKumar2121** [comment]
In Node.js, the choice between Buffer and TypedArrays depends on your needs. Buffer has a bigger API and was re-implemented for compatibility and performance. Performance may vary, so benchmark your specific use case. If you want a larger API and are comfortable with potential quirks, go with Buffer. If you prefer a standardized and predictable API, use TypedArrays like Uint8Array, following Node.js documentation recommendations.
